qualtran/bloqs/data_loading/qroam_clean.py

```diff
@@ -0,0 +1,389 @@
+
+    @classmethod
+    def build_from_bitsize(
+        cls: Type['QROAMClean'],
+        data_len_or_shape: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        target_bitsizes: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        *,
+        selection_bitsizes: Tuple[SymbolicInt, ...] = (),
+        num_controls: SymbolicInt = 0,
+        log_block_sizes: Optional[Union[SymbolicInt, Tuple[SymbolicInt, ...]]] = None,
+    ) -> 'QROAMClean':
+        qroam: 'QROAMClean' = cls._build_from_bitsize(
+            data_len_or_shape,
+            target_bitsizes,
+            selection_bitsizes=selection_bitsizes,
+            num_controls=num_controls,
+        )
+        return qroam.with_log_block_sizes(log_block_sizes=log_block_sizes)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                *self.control_registers,
+                *self.selection_registers,
+                *self.target_registers,
+                *self.junk_registers,
+            ]
+        )
+
+    @cached_property
+    def batched_data_permuted(self) -> List[np.ndarray]:
+        if is_symbolic(*self.block_sizes):
+            raise ValueError(
+                f"Cannot decompose SelectSwapQROM bloq with symbolic block sizes. Found {self.block_sizes=}"
+            )
+        block_sizes = cast(Tuple[int, ...], self.block_sizes)
+        ret = []
+        for data, swz in zip(self.batched_data, self.swap_with_zero_bloqs):
+            permuted_batched_data = np.zeros(data.shape + block_sizes, dtype=data.dtype)
+            for sel_l in np.ndindex(cast(Tuple[int, ...], self.batched_qrom_shape)):
+                for sel_k in np.ndindex(block_sizes):
+                    sel_kwargs = {reg.name: sel for reg, sel in zip(swz.selection_registers, sel_k)}
+                    curr_data = swz.call_classically(**sel_kwargs, targets=np.copy(data[sel_l]))[-1]
+                    idx = (*sel_l, *sel_k)
+                    permuted_batched_data[idx][:] = curr_data
+
+            n_blocks = len(self.block_sizes)
+            transpose_axes = [x for i in range(n_blocks) for x in [i, i + n_blocks]]
+            transpose_axes += [i + 2 * n_blocks for i in range(n_blocks)]
+            desired_shape = tuple(
+                NbyK * K for NbyK, K in zip(self.batched_qrom_shape, self.block_sizes)
+            )
+            ret.append(
+                permuted_batched_data.transpose(transpose_axes).reshape(
+                    desired_shape + self.block_sizes
+                )
+            )
+        return ret
+
+    @cached_property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        # The newly allocated registers should be kept around for measurement based uncomputation.
+        junk_regs = []
+        block_size = prod(self.block_sizes)
+        for reg in self.target_registers:
+            assert reg.shape == ()
+            if is_symbolic(block_size) or block_size > 1:
```

Comment:
Can you still get a saving for QROM inverse when block_size == 1 (regular old QROM)? I have a feeling you can but maybe that should be handled elsewhere?

Reply:
Yes, it's already handled. You just pass log_block_size=0 to the `QROAMClean` bloq. 

<img width="1237" alt="image" src="https://github.com/user-attachments/assets/0c07b3d1-e5cd-411f-8699-973db2fbd94d">

Reply:
Awesome.